File: lexer.py
from sys import exit

TOKEN_TYPE = {
    "Word": "Word",
    "Number": "Number",
    "String": "String",
    "Array": "Array",
    "Dict": "Dict",
    "Func": "Func",
    "CallFunc": "CallFunc",
    "Var": "Var",
    "If": "If",
    "ElseIf": "ElseIf",
    "Else": "Else",
    "While": "While",
    "For": "For",
    "True": "True",
    "False": "False",
    "And": "And",
    "Or": "Or",
    "Return": "Return",
    "Import": "Import",
    "Export": "Export",
    "Module": "Module",
    "LeftParen": "LeftParen",
    "RightParen": "RightParen",
    "LeftBrace": "LeftBrace",
    "RightBrace": "RightBrace",
    "LeftBracket": "LeftBracket",
    "RightBracket": "RightBracket",
    "Comma": "Comma",
    "Plus": "Plus",
    "Minus": "Minus",
    "Times": "Times",
    "Divide": "Divide",
    "Equal": "Equal",
    "Equality": "Equality",
    "LessThan": "LessThan",
    "GreaterThan": "GreaterThan",
    "Eof": "Eof",
    "Class": "Class",
    "Constructor": "Constructor",
    "Self": "Self",
    "New": "New",
    "Range": "Range"
}

KEYWORDS = {
    "being": TOKEN_TYPE["Import"],
    "containing": TOKEN_TYPE["Module"],
    "burn": TOKEN_TYPE["Var"],
    "for": TOKEN_TYPE["For"],
    "through": TOKEN_TYPE["Range"],
    "prepmatch": TOKEN_TYPE["Func"],
    "while": TOKEN_TYPE["While"],
    "lightmatch": TOKEN_TYPE["CallFunc"],
    "lightertype": TOKEN_TYPE["Class"],
    "ignite": TOKEN_TYPE["Constructor"],
    "self": TOKEN_TYPE["Self"],
    "pullout": TOKEN_TYPE["New"],
    "leave": TOKEN_TYPE["Export"],
    "if": TOKEN_TYPE["If"],
    "elif": TOKEN_TYPE["ElseIf"],
    "else": TOKEN_TYPE["Else"]
}

def new_token(kind, value, content):
    return {
        "type": kind,
        "value": value,
        "content": content
    }

class Lexer:
    def __init__(self, source="", current=0, tokens=[], line=0):
        self.current = current
        self.source = source
        self.tokens = tokens
        self.line = line

    def peek(self):
        if self.current >= len(self.source): return "\0"
        return self.source[self.current]

    def peek_next(self):
        if self.current >= len(self.source): return "\0"
        return self.source[self.current + 1]

    def advance(self):
        if self.current >= len(self.source): return "\0"
        res = self.peek()
        self.current += 1
        return res

    def match(self, char):
        if self.peek() == char:
            self.advance()
            return True
        return False
    
    def add_token(self, kind, value, content):
        self.tokens.append(new_token(kind, value, content))

def scan_token(lexer):
    char = lexer.advance()

    def is_alphanumeric(char):
        return char != " " and (char.isalpha() or char.isnumeric() or char == "_")

    def string(kind):
        text = ""
        while lexer.peek() != kind and lexer.peek() != "\0":
            if lexer.peek() == "\n":
                lexer.line += 1
            text += lexer.advance()
        if lexer.peek() == "\0":
            # Reached end of file, but string hasn't been terminated
            raise Exception(f"Unterminated string: {lexer.line}")
        lexer.advance()  # Consume the closing quote
        lexer.add_token(TOKEN_TYPE["String"], text, text)

    def number():
        text = ""
        while lexer.peek().isnumeric():
            text += lexer.advance()
        if lexer.peek() == "." and lexer.peek_next().isnumeric():
            text += lexer.advance()
            while lexer.peek().isnumeric():
                text += lexer.advance()
        lexer.add_token(TOKEN_TYPE["Number"], float(text), text)

    def identifier():
        text = ""
        while is_alphanumeric(lexer.peek()):
            text += lexer.advance()
        kind = KEYWORDS.get(text, None)
        if kind is None:
            kind = TOKEN_TYPE["Word"]
        lexer.add_token(kind, text, text)

    match char:
        case "(":
            lexer.add_token(TOKEN_TYPE["LeftParen"], "(", "(")
        case ")":
            lexer.add_token(TOKEN_TYPE["RightParen"], ")", ")")
        case "{":
            lexer.add_token(TOKEN_TYPE["LeftBrace"], "{", "{")
        case "}":
            lexer.add_token(TOKEN_TYPE["RightBrace"], "}", "}")
        case "[":
            lexer.add_token(TOKEN_TYPE["LeftBracket"], "[", "[")
        case "]":
            lexer.add_token(TOKEN_TYPE["RightBracket"], "]", "]")
        case ",":
            lexer.add_token(TOKEN_TYPE["Comma"], ",", ",")
        case "+":
            lexer.add_token(TOKEN_TYPE["Plus"], "+", "+")
        case "-":
            lexer.add_token(TOKEN_TYPE["Minus"], "-", "-")
        case "*":
            lexer.add_token(TOKEN_TYPE["Times"], "*", "*")
        case "/":
            lexer.add_token(TOKEN_TYPE["Divide"], "/", "/")
        case "=":
            lexer.add_token(TOKEN_TYPE["Equal"], "=", "=")
        case '"':
            string('"')
        case "'":
            string("'")
        case "<":
            lexer.add_token(TOKEN_TYPE["LessThan"], "<", "<")
        case ">":
            lexer.add_token(TOKEN_TYPE["GreaterThan"], ">", ">")
        case "#":
            while lexer.peek() != "\n":
                lexer.advance()
            lexer.line += 1
        case "\n":
            lexer.line += 1
        case " ":
            return
        case _:
            if char.isalpha():
                lexer.current -= 1
                identifier()
            elif char.isnumeric():
                lexer.current -= 1
                number()
            else:
                raise Exception(f"Unexpected character: {char} at {lexer.line + 1}:{lexer.current + 1}")

def scan_tokens(lexer):
    # Main wrapper function
    # Let's scan the tokens one by one, and place them in lexer.tokens
    while lexer.current < len(lexer.source):
       scan_token(lexer)
    lexer.add_token(TOKEN_TYPE["Eof"], "", "")
    return lexer.tokens

File: src/lexer.rs
use std::collections::HashMap;

#[derive(Clone, Debug, PartialEq)]
pub enum TokenType {
    Word,
    Number,
    String,
    Array,
    Dict,
    Func,
    Call,
    Var,
    If,
    Elif,
    Else,
    While,
    For,
    True,
    False,
    And,
    Or,
    Return,
    LeftParen,
    RightParen,
    LeftBrace,
    RightBrace,
    LeftBracket,
    RightBracket,
    Comma,
    Plus,
    Minus,
    Times,
    Divide,
    Modulo,
    Equal,
    Equality,
    LessThan,
    LessThanOrEqual,
    GreaterThan,
    GreaterThanOrEqual,
    Eof,
    Class,
    New,
    Range,
    Colon,
    Period,
}

pub fn keywords() -> HashMap<&'static str, TokenType> {
    let mut keywords = HashMap::new();
    keywords.insert("return", TokenType::Return);
    keywords.insert("burn", TokenType::Var);
    keywords.insert("for", TokenType::For);
    keywords.insert("through", TokenType::Range);
    keywords.insert("prepmatch", TokenType::Func);
    keywords.insert("while", TokenType::While);
    keywords.insert("lightertype", TokenType::Class);
    keywords.insert("pullout", TokenType::New);
    keywords.insert("if", TokenType::If);
    keywords.insert("elif", TokenType::Elif);
    keywords.insert("else", TokenType::Else);
    keywords.insert("True", TokenType::True);
    keywords.insert("False", TokenType::False);
    keywords.insert("and", TokenType::And);
    keywords.insert("or", TokenType::Or);
    return keywords;
}

#[derive(Debug)]
pub struct Token {
    pub kind: TokenType,
    pub value: String,
    pub content: String,
}

impl Token {
    pub fn new(kind: TokenType, value: String, content: String) -> Token {
        Token {
            kind,
            value,
            content,
        }
    }
}

pub struct Lexer {
    source: String,
    pub tokens: Vec<Token>,
    current: usize,
    line: usize,
}

fn char_at(s: &String, index: usize) -> Option<char> {
    s.chars().nth(index)
}

impl Lexer {
    pub fn new(source: String) -> Lexer {
        Lexer {
            source,
            tokens: Vec::new(),
            current: 0,
            line: 0,
        }
    }

    pub fn peek(&self) -> char {
        if self.current >= self.source.len() {
            return '\0';
        }
        char_at(&self.source, self.current).unwrap()
    }

    pub fn peek_next(&self) -> char {
        if self.current >= self.source.len() {
            return '\0';
        }
        char_at(&self.source, self.current + 1).unwrap()
    }

    pub fn advance(&mut self) -> char {
        if self.current >= self.source.len() {
            return '\0';
        }
        let res = self.peek();
        self.current += 1;
        res
    }

    pub fn r#match(&mut self, chr: char) -> bool {
        if self.peek() == chr {
            self.advance();
            return true;
        }
        false
    }

    pub fn add_token(&mut self, kind: TokenType, value: String, content: String) {
        self.tokens.push(Token::new(kind, value, content))
    }
}

pub fn scan_token(lexer: &mut Lexer) {
    let chr = lexer.advance();

    fn string(kind: char, lexer: &mut Lexer) {
        let mut text = String::new();
        while lexer.peek() != kind && lexer.peek() != '\0' {
            if lexer.peek() == '\n' {
                lexer.line += 1;
            }
            text.push_str(&lexer.advance().to_string());
        }
        if lexer.peek() == '\0' {
            // Reached end of file, but string hasn't been terminated
            panic!("Unterminated string: {}", lexer.line);
        }
        lexer.advance(); // Consume the closing quote
        lexer.add_token(TokenType::String, text.clone(), text.clone());
    }

    fn number(lexer: &mut Lexer) {
        let mut text = String::new();
        while lexer.peek().is_numeric() {
            text.push(lexer.advance());
        }
        if lexer.peek() == '.' && lexer.peek_next().is_numeric() {
            text.push(lexer.advance());
            while lexer.peek().is_numeric() {
                text.push(lexer.advance());
            }
        }
        lexer.add_token(TokenType::Number, text.clone(), text.clone());
    }

    fn identifier(lexer: &mut Lexer) {
        let mut text = String::new();
        while lexer.peek().is_alphabetic() {
            text.push(lexer.advance());
        }
        match keywords().get(&text.as_str()) {
            Some(value) => {
                lexer.add_token(value.clone(), text.clone(), text.clone());
            }
            None => {
                lexer.add_token(TokenType::Word, text.clone(), text.clone());
            }
        }
    }

    match chr {
        '(' => lexer.add_token(TokenType::LeftParen, String::from("("), String::from("(")),
        ')' => lexer.add_token(TokenType::RightParen, String::from(")"), String::from(")")),
        '{' => lexer.add_token(TokenType::LeftBrace, String::from("{"), String::from("{")),
        '}' => lexer.add_token(TokenType::RightBrace, String::from("}"), String::from("}")),
        '[' => lexer.add_token(TokenType::LeftBracket, String::from("["), String::from("[")),
        ']' => lexer.add_token(
            TokenType::RightBracket,
            String::from("]"),
            String::from("]"),
        ),
        ',' => lexer.add_token(TokenType::Comma, String::from(","), String::from(",")),
        '+' => lexer.add_token(TokenType::Plus, String::from("+"), String::from("+")),
        '-' => lexer.add_token(TokenType::Minus, String::from("-"), String::from("-")),
        '*' => lexer.add_token(TokenType::Times, String::from("*"), String::from("*")),
        '/' => lexer.add_token(TokenType::Divide, String::from("/"), String::from("/")),
        '%' => lexer.add_token(TokenType::Modulo, String::from("%"), String::from("%")),
        '=' => {
            if lexer.peek() == '=' {
                lexer.advance();
                lexer.add_token(
                    TokenType::LessThanOrEqual,
                    String::from("<="),
                    String::from("<="),
                );
            } else {
                lexer.add_token(TokenType::LessThan, String::from("<"), String::from("<"));
            }
        }
        '"' => string('"', lexer),
        '\'' => string('\'', lexer),
        '<' => {
            if lexer.peek() == '=' {
                lexer.advance();
                lexer.add_token(
                    TokenType::LessThanOrEqual,
                    String::from("<="),
                    String::from("<="),
                );
            } else {
                lexer.add_token(TokenType::LessThan, String::from("<"), String::from("<"));
            }
        }
        '>' => {
            if lexer.peek() == '=' {
                lexer.advance();
                lexer.add_token(
                    TokenType::GreaterThanOrEqual,
                    String::from(">="),
                    String::from(">="),
                );
            } else {
                lexer.add_token(TokenType::GreaterThan, String::from(">"), String::from(">"));
            }
        }
        ':' => lexer.add_token(TokenType::Colon, String::from(":"), String::from(":")),
        '.' => lexer.add_token(TokenType::Period, String::from("."), String::from(".")),
        '#' => {
            while lexer.peek() != '\n' && lexer.peek() != '\0' {
                lexer.advance();
            }
            lexer.line += 1;
        }
        '\n' => {
            lexer.line += 1;
        }
        ' ' => {
            return;
        }
        '\t' => {
            return;
        }
        _ => {
            if chr.is_alphabetic() {
                lexer.current -= 1;
                identifier(lexer);
            } else if chr.is_numeric() {
                lexer.current -= 1;
                number(lexer);
            } else {
                panic!("Unexpected character: {} at line {}", chr, lexer.line + 1);
            }
        }
    }
}

pub fn scan_tokens(lexer: &mut Lexer) -> &Vec<Token> {
    while lexer.current < lexer.source.len() {
        scan_token(lexer)
    }
    lexer.add_token(TokenType::Eof, String::new(), String::new());
    &lexer.tokens
}
